src/transform/src/redundant_join.rs

```diff
@@ -733,3 +769,82 @@ fn try_build_expression_using_other(
         }
     }
 }
+
+/// A context of `ProvInfo` vectors associated with bindings that might still be
+/// referenced.
+#[derive(Debug, Default)]
+pub struct ProvInfoCtx {
+    /// [`LocalId`] references in the remaining subtree.
+    ///
+    /// Entries from the `lets` map that are no longer used can be pruned.
+    uses: BTreeMap<LocalId, usize>,
+    /// [`ProvInfo`] vectors associated with let binding in scope.
+    lets: BTreeMap<LocalId, Vec<ProvInfo>>,
+}
+
+impl ProvInfoCtx {
+    /// Extend the `uses` map by the `LocalId`s used in `expr` and remove
+    /// `lets` entries for `uses` that reset to zero.
+    pub fn extend_uses(&mut self, expr: &MirRelationExpr) {
+        expr.visit_pre(&mut |expr: &MirRelationExpr| match expr {
+            MirRelationExpr::Get {
+                id: Id::Local(id),
+                typ: _,
+            } => {
+                let count = self.uses.entry(id.clone()).or_insert(0_usize);
+                *count += 1;
+            }
+            _ => (),
+        });
+    }
+
+    /// Decrement `uses` entries by the `LocalId`s used in `expr` and remove
+    /// `lets` entries for `uses` that reset to zero.
+    pub fn remove_uses(&mut self, expr: &MirRelationExpr) {
+        let mut worklist = vec![expr];
+        while let Some(expr) = worklist.pop() {
+            if let MirRelationExpr::Get {
+                id: Id::Local(id), ..
+            } = expr
+            {
+                if let Some(count) = self.uses.get_mut(id) {
+                    if *count > 0 {
+                        *count -= 1;
+                    }
+                    if *count == 0 {
+                        self.lets.remove(id);
+                    }
+                }
```

Comment:
This was not strictly speaking a bug, but a by-product of where `ProvInfoCtx::remove_uses` was called and at what sub-terms it was not descending.

Prior to the latest revision:

1. `ProvInfoCtx::remove_uses` was called after visiting a `Let` / `LetRec` value.
2. The visitation in the `remove_uses` body was not descending in the `values` of a `Let` / `LetRec` node, but was descending in the body.
3. However, the `Let` / `LetRec` case in `RedundantJoin::action` was removing the `ctx` entry for locally IDs before leaving the case. 

Because of this, in situations such as this:

```text
Return
  Get l1
With Mutually Recursive
  cte l1 =
    Reduce group_by=[#0] aggregates=[min(#1)]
      Union
        Project (#1, #1)
          Get y
        Project (#1, #3)
          Join on=(#0 = #2)
            Filter (#0) IS NOT NULL
              Get l0
            Filter (#0) IS NOT NULL
              Get l1
```

When calling `remove_uses` from the context in which the whole expression was living, the `Get l1` was already removed.

I changed this to

1. Call `ProvInfoCtx::remove_uses` after visiting a `Let` / `LetRec` value **and** after the body.
2. The visitation in the `remove_uses` body now is not descending `Let` / `LetRec` nodes altogether.